content/oci/deletable.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package oci provides access to an OCI content store.
+// Reference: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc4/image-layout.md
+package oci
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+)
+
+// DeletableStore implements `oras.Target`, and represents a content store
+// extended with the delete operation.
+// Reference: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc4/image-layout.md
+type DeletableStore struct {
+	*Store
+	lock sync.RWMutex
+}
+
+// NewDeletableStore returns a new DeletableStore with context.Background().
+func NewDeletableStore(root string) (*DeletableStore, error) {
+	return NewDeletableStoreWithContext(context.Background(), root)
+}
+
+// NewDeletableStoreWithContext returns a new DeletableStore.
+func NewDeletableStoreWithContext(ctx context.Context, root string) (*DeletableStore, error) {
+	store, err := NewWithContext(ctx, root)
+	if err != nil {
+		return nil, err
+	}
+	return &DeletableStore{
+		Store: store,
+	}, nil
+}
+
+// Fetch fetches the content identified by the descriptor. It returns an io.ReadCloser.
+// It's recommended to close the io.ReadCloser before a DeletableStore.Delete
+// operation, otherwise Delete may fail on some systems.
+func (ds *DeletableStore) Fetch(ctx context.Context, target ocispec.Descriptor) (io.ReadCloser, error) {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	return ds.Store.Fetch(ctx, target)
+}
+
+// Push pushes the content, matching the expected descriptor. This implementation
+// does not allow concurrent Push. To use concurrent Push, consider Store.Push.
+func (ds *DeletableStore) Push(ctx context.Context, expected ocispec.Descriptor, reader io.Reader) error {
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	return ds.Store.Push(ctx, expected, reader)
+}
+
+// Delete removes the content matching the descriptor from the store. Delete may
```

Comment:
Delete and remove are different wording. Consider "remove a song from a playlist" and "delete a song from the music library". Basically, delete contents and remove references.
```suggestion
// Delete deletes the content matching the descriptor from the store. Delete may
```

Reply:
changed accordingly.